File: Go/Best Time to Buy and Sell Stock II.go
// 4ms, 92.66%; 3.1MB, 100.00%
func maxProfit(prices []int) int {
	res, minPrice, length := 0, 0x7fffffff, len(prices)
	for i := 0; i < length; {
		for ; i < length && prices[i] <= minPrice; i++ {
			minPrice = prices[i]
		}
		if i == length {
			break
		}
		for ; i < length && prices[i] >= prices[i-1]; i++ {
		}
		res += prices[i-1] - minPrice
		minPrice = 0x7fffffff
	}
	return res
}

File: Java/Best Time to Buy and Sell Stock II.java
// 0ms, 100.00%; 39.7MB, 12.38%
class Solution {
    public int maxProfit(int[] prices) {
        int res = 0, beg, end = -1;
        while (true) {
            beg = ascendUntil(prices, end + 1, false);
            end = ascendUntil(prices, beg, true);
            if (end == prices.length) {
                break;
            }
            res += prices[end] - prices[beg];
        }
        return res;
    }

    private static int ascendUntil(int[] arr, int beg, boolean isAscend) {
        int ret;
        for (ret = beg + 1; ret < arr.length; ++ret) {
            if (isAscend == (arr[ret - 1] > arr[ret])) {
                break;
            }
        }
        return ret - 1;
    }
}
